lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2016,6 +2023,85 @@ postTransactionFeeOld ctx (ApiT wid) body = do
         minCoins <- liftIO (W.calcMinimumCoinValues @_ @k wrk outs)
         liftHandler $ mkApiFee Nothing minCoins <$> W.estimateFee runSelection
 
+balanceTransaction
+    :: forall ctx s k (n :: NetworkDiscriminant).
+        ( ctx ~ ApiLayer s k
+        , IsOwned s k
+        , WalletKey k
+        , GetRewardAccount s k
+        , HardDerivation k
+        , GenChange s
+        , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+        )
+    => Proxy n
+    -> ctx
+    -> ConstructTransactionConfig s IO
+    -> ApiT WalletId
+    -> ApiBalanceTransactionPostData n
+    -> Handler (ApiConstructTransaction n)
+balanceTransaction _ ctx config (ApiT wid) body =
+    if areOutputsCovered tx then
+        liftHandler $ throwE ErrBalanceTxAllOutputsCovered
+    else do
+        ApiConstructTransaction _ cs _ <-
+            constructTransaction ctx config (ApiT wid) $ toApiConstructTransactionData tx
+        (sealedTxOutcoming, newfee) <-
+            case updateTx tl sealedTxIncoming (getInpsChange cs) of
+                Right result -> pure result
+                Left err ->
+                    liftHandler $ throwE $ ErrBalanceTxUpdateError err
+        pure $ ApiConstructTransaction
+            { transaction = ApiT sealedTxOutcoming
+            , coinSelection = cs
+            , fee = Quantity $ fromIntegral newfee
+            }
+  where
+    toApiConstructTransactionData (Tx _ _ _ outs wdrlM mdM) = ApiConstructTransactionData
+        { payments = toApiPaymentDesination (snd $ outsAdjusted outs)
+        , withdrawal = toApiWdrl wdrlM
+        , metadata = ApiT <$> mdM
+        , mint = Nothing
+        , delegations = Nothing
+        , validityInterval = Nothing
+        }
+    toApiWdrl :: Map RewardAccount Coin -> Maybe ApiWithdrawalPostData
+    toApiWdrl _ = Nothing --TODO in ADP-656 - I need to lookup the wallet rewardAcct against this map?
+    toApiPaymentDesination = \case
+        [] -> Nothing
+        txouts -> Just $ ApiPaymentAddresses $ NE.fromList $ toAddressAmount <$> txouts
+    toAddressAmount :: TxOut -> AddressAmount (ApiT Address, Proxy n)
+    toAddressAmount (TxOut addr (TokenBundle coin tokenMap)) = AddressAmount
+        { address = (ApiT addr, Proxy)
+        , amount = Quantity $ fromIntegral (unCoin coin)
+        , assets = ApiT tokenMap
+        }
+
+    toTxIn (ApiCoinSelectionInput (ApiT txid) ix _ _ _ _) =
+        TxIn txid ix
+    toTxOut (ApiCoinSelectionChange (ApiT addr,_) (Quantity amt) (ApiT assets) _) =
+        TxOut addr (TokenBundle (Coin $ fromIntegral amt) assets)
+    getInpsChange cs =
+        ( map toTxIn $ NE.toList $ cs ^. #inputs
+        , map toTxOut $ cs ^. #change)
+
+    areOutputsCovered (Tx _ feeM _ outs _ _) = -- TODO in ADP-656 is it correct- double check
```

Comment:
I think there's value in moving meaningful logic outside the handler. For example, this concept of "Is this transaction balanced" could be useful to other modules, and is more easily tested outside the handler. Inside the handler, it's very difficult to test this logic.